src/search/successor_generators/yannakakis.h

```diff
@@ -15,11 +17,41 @@ class YannakakisSuccessorGenerator : public GenericJoinSuccessor {
 
  private:
   std::vector<std::vector<std::pair<int, int>>> full_reducer_order;
-  std::vector<std::vector<std::pair<int, int>>> join_tree_order;
-  std::vector<std::vector<int>> remaining_join;
-  std::vector<std::vector<int>> number_of_child;
+
   std::vector<std::unordered_set<int>> distinguished_variables;
+
+  std::vector<std::vector<int>> remaining_join;
+
+  std::vector<JoinTree> join_trees;
+
   void get_distinguished_variables(const ActionSchema &action);
 };
 
+class JoinTree {
+    std::vector<std::pair<int, int>> join_tree_order;
+    std::vector<int> number_of_children;
+
+public:
+    JoinTree() = default;
+
+    void add_node(int i, int j) {
+        join_tree_order.emplace_back(i, j);
+        number_of_children[j]++;
+    }
+
+    void set_number_of_nodes(size_t i) {
+        number_of_children.resize(i, 0);
+    }
+
+    std::vector<int> get_copy_number_children() const {
```

Comment:
I wouldn't name the methods with "copy" or non-copy, that can be deduced from the signature. Also, better to return a const ref to the vector, if caller wants a copy, she can call with auto x = jt.get_number_children(), and that makes the copy. This way, the method is useful for possible callers not wanting a copy.